### ISSUE TITLE
no-use-before-define issue

### ISSUE THREAD

**Issue by rohanorton** [eslint/eslint#4280]
Hi, I'm new to eslint so there might be something that I'm missing, but I was expecting the rule "no-use-before-define" to pick up the following (contrived) syntax issue:

... | ....
----|----
version | v1.7.3
rule | no-use-before-define
code | see below
expected | output error
actual | no output

code:
```js
var a = a
```

eslintrc:
```json
{
    "extends": "eslint:recommended",
    "rules": {
         "no-use-before-define": 2
    }
}
```

**Comment by eslintbot** [comment]
Thanks for the issue! If you're reporting a bug, please be sure to include:

1. The version of ESLint you are using (run `eslint -v`)
2. What you did (the source code and ESLint configuration)
3. The actual ESLint output complete with numbers
4. What you expected to happen instead

Requesting a new rule? Please see [Proposing a New Rule](http://eslint.org/docs/developer-guide/contributing/new-rules) for instructions.

**Comment by ilyavolodin** [comment]
Hmm.. That's an interesting case. I think the code above is valid, and will be resolved as:
```js
var a;
a = a;
```

I'm not 100% sure that it should be marked as an issue by this rule it feels more in line with `no-self-compare`. @eslint/eslint-team

**Comment by mysticatea** [comment]
```js
let a = a; // this second `a` is in the TDZ.
```

I guess this pattern is going to throw a runtime error, so I think there is a reason that this rule should report this pattern.

![image](https://cloud.githubusercontent.com/assets/1937871/10803357/25e9414a-7e05-11e5-9c03-661e25798e12.png)

**Comment by nzakas** [comment]
That's a good point. Let's add it.

**Comment by unional** [comment]
Another issue:

````
var b = function() {
    a = 2;
};

var a = {
    x: b
};
````

**Comment by nzakas** [comment]
@unional what's the issue?

**Comment by unional** [comment]
Circular reference.  
````
var b = function () {
  a = 2; // no-used-before-defined
};
var a = {
  x: b
};
````

**Comment by unional** [comment]
Here is a more liberate example:

````
var x = {
  a: {
    fn: function() {
        $.post('sofa', null, y.someCallback.bind(y));
        y.doSomething();
    }
  }
};

var y = {
  b: [x]
  someCallback: () => {...},
  doSomething: () => {....}
};
````

**Comment by nzakas** [comment]
This is working as expected. You need to disable the rule to allow those patterns. Remember, this rule is not about when something is safe to use, is about whether something is defined prior to its use in your source code.

**Comment by mysticatea** [comment]
I will work on this after #4597 is merged.

**Comment by mysticatea** [comment]
lol

I found several potential bugs in ESLint by this enhancement.

**Comment by unional** [comment]
Nice! :+1:
